### DISCUSSION TITLE
What is the 'correct' way to use Pocketbase with a SPA

### DISCUSSION THREAD

**Discussion by AllanOcelot** [pocketbase/pocketbase#2046] [Q&A] [answered]
Hi all, 

First of all apologies for the 'noob' question - I've been playing around with pocketbase and absolutely love it. It does everything I need and the JS SDK is great. I have zero complaints and plan to stick around should I be able to contribute to the frontend in any capacity. 

That being said  - I'm a career frontend developer, the last time I did any 'heavy lifting' with backend systems would have been with PHP years back.

Currently I have a VueJS SPA that connects to a Node app, the node app then queries Pocketbase, and returns the data to the SPA. 

**Is this the 'correct' way to implement Pocketbase in an SPA**? 
I'm sure I can just use the JS SDK via vue and make API calls bypassing node - I just want to know what the 'correct' implementation is.  I've tried to research this but I'm hitting a brick wall. My main concern here is security, followed by 'doing it right'.

Thank you for your time, and long live Pocketbase!

**Comment by ganigeorgiev** [comment, marked as answer]
This is difficult to generalize and I don't think there is one way of "correct" and "doing it right", especially when integration with multiple applications and tools is involved.

It really depends on your use case, with what tools you feel comfortable working with, and whether you want to implement an additional server-side business logic (eg. performing custom data validations, handling payment processing webhooks, etc.).

If you don't have a need for a custom server-side handling OR you [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) to implement your custom business logic, then the client-side application can interact directly with the PocketBase server without a middleware service (_this is my preferred way of working with PocketBase as it was designed for this exact case, see [Presentator #183](https://github.com/presentator/presentator/issues/183)_).

If you are not feeling comfortable working with Go (_note that JS app event hooks are in the [roadmap](https://github.com/orgs/pocketbase/projects/2)_) and you need SSR or custom server-side requests handling, then you can implement a middleware service (in your case with Node.js) that will interact with PocketBase behind the scenes.

When using a middleware service, you usually have 2 options:

- connect directly from the client-side to PocketBase for all basic record operations and use the node service only for handling the custom server-side part of your application (_the [JS SDK](https://github.com/pocketbase/js-sdk) can be used for both regular user and admin level actions_).

- hide PocketBase entirely behind the Node service, meaning that the clients will interact directly only with your Node application, which could perform some request data transformations and then send the appropriate requests back to the PocketBase server.

The latter is similar to the example SSR integration shown in https://github.com/pocketbase/js-sdk/tree/master#ssr-integration. It uses a cookie based flow to share the auth state in client<->node<->PocketBase. If you need the mixed approach (aka. both client and node can interact with PocketBase), the flow should be more-or-less the same with the small difference that on the client-side you need to read and set the cookie using `document.cookie` in addition to the node request and response headers object (_some frameworks have cookies abstraction that could help with this_).

**Reply by AllanOcelot** [reply]
Perfect, you've answered all my concerns - thanks for your time. 

I'll stick with the Node route for now as I've previous experience with this setup and firebase, and it feels 'right' to hide pocketbase.  Plus I do plan on adding more logic there down the line.

Excited to hear about the JS event hooks!
